### PR TITLE
SLVS-1114 "NaN" should not be used in comparisons.

### PR DESCRIPTION
**Pull request by valhristov**

### CODE REVIEW

File: src/SonarAnalyzer.CSharp/Rules/CompareNaN.cs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+                return false;
+            }
+
+            var fieldSymbol = semanticModel.GetSymbolInfo(memberAccess).Symbol as IFieldSymbol;
+            if (fieldSymbol?.Name != nameof(double.NaN))
+            {
+                return false;
+            }
+
+            floatingPointType = KnownType.FloatingPointNumbers.FirstOrDefault(fieldSymbol.Type.Is);
```
Comment (SonarTech) [inline src/SonarAnalyzer.CSharp/Rules/CompareNaN.cs]: BLOCKER "fieldSymbol" is null on at least one execution path. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS2259)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* BLOCKER 1 blocker

Watch the comments in this conversation to review them.

File: src/SonarAnalyzer.CSharp/Rules/CompareNaN.cs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+            var memberAccess = expression as MemberAccessExpressionSyntax;
+            if (memberAccess == null)
+            {
+                return false;
+            }
+
+            var fieldSymbol = semanticModel.GetSymbolInfo(memberAccess).Symbol as IFieldSymbol;
+            if (fieldSymbol?.Name == nameof(double.NaN))
+            {
+                floatingPointType = KnownType.FloatingPointNumbers.FirstOrDefault(fieldSymbol.Type.Is);
```
Comment (SonarTech) [inline src/SonarAnalyzer.CSharp/Rules/CompareNaN.cs]: BLOCKER "fieldSymbol" is null on at least one execution path. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS2259)

File: src/SonarAnalyzer.CSharp/Rules/CompareNaN.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            var memberAccess = expression as MemberAccessExpressionSyntax;
+            if (memberAccess == null)
+            {
+                return false;
+            }
+
+            var fieldSymbol = semanticModel.GetSymbolInfo(memberAccess).Symbol as IFieldSymbol;
+            if (fieldSymbol?.Name == NameOfNaN)
+            {
+                floatingPointType = KnownType.FloatingPointNumbers.FirstOrDefault(fieldSymbol.Type.Is);
```
Comment (SonarTech) [inline src/SonarAnalyzer.CSharp/Rules/CompareNaN.cs]: BLOCKER "fieldSymbol" is null on at least one execution path. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS2259)

File: src/SonarAnalyzer.CSharp/Rules/CompareNaN.cs [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            floatingPointType = null;
+
+            var memberAccess = expression as MemberAccessExpressionSyntax;
+            if (memberAccess == null)
+            {
+                return false;
+            }
+
+            var fieldSymbol = semanticModel.GetSymbolInfo(memberAccess).Symbol as IFieldSymbol;
+            if (fieldSymbol != null && fieldSymbol.Name == nameof(double.NaN))
```
Comment (Evangelink) [inline src/SonarAnalyzer.CSharp/Rules/CompareNaN.cs]: I think it is worth adding a comment here to mention you are testing to null just to avoid a False Positive. Also you should create the associated ticket.

Comment (tvsonar) [inline src/SonarAnalyzer.CSharp/Rules/CompareNaN.cs]: @valhristov Did you create a ticket for this issue? Namely, we should encode the fact that `nameof` never returns `null`. Also, I wouldn't change the code because of false positives. Instead the false positive should be fixed. (cc @Evangelink)
